src/internal/attr.c

```diff
@@ -0,0 +1,189 @@
+#include <rlang.h>
+#include "internal.h"
+
+static inline sexp* r_language_names(sexp* x);
+static inline sexp* r_names_dispatch(sexp* x);
+
+sexp* rlang_names2(sexp* x) {
+  const enum r_type type = r_typeof(x);
+
+  if (type == r_type_environment) {
+    r_abort("Use `env_names()` for environments.");
+  }
+
+  // Handle pairlists and language objects specially like `getAttrib()`
+  // does. `r_names()` will not find these names because it has a guarantee
+  // to never allocate.
+  if (type == r_type_pairlist || type == r_type_call) {
+    return r_language_names(x);
+  }
+
+  sexp* nms;
+  if (r_is_object(x)) {
+    nms = KEEP(r_names_dispatch(x));
+  } else {
+    nms = KEEP(r_names(x));
+  }
+
+  if (r_is_null(nms)) {
+    r_ssize n = r_length(x);
+    nms = KEEP(r_new_vector(r_type_character, n));
+    r_chr_fill(nms, r_empty_str, n);
+  } else {
+    nms = KEEP(rlang_replace_na(nms, r_shared_empty_chr));
+  }
+
+  FREE(2);
+  return nms;
+}
+
+static inline sexp* r_language_names(sexp* x) {
+  r_ssize n = r_length(x);
+
+  sexp* out = KEEP(r_new_vector(r_type_character, n));
+  sexp** p_out = STRING_PTR(out);
+
+  int i = 0;
+
+  for(; x != r_null; x = r_node_cdr(x), ++i) {
+    sexp* tag = r_node_tag(x);
+
+    if (tag == r_null) {
+      p_out[i] = r_empty_str;
+    } else if (r_is_symbolic(tag)) {
+      p_out[i] = PRINTNAME(tag);
+    } else {
+      r_abort("Invalid type %s for tag", r_type_as_c_string(r_typeof(tag)));
+    }
+  }
+
+  FREE(1);
+  r_mark_shared(out);
```

Comment:
Why is this marked shared?

Reply:
Same reasoning as above, looking at what base R does
https://github.com/wch/r-source/blob/3d8c9fcaa927eedf0ec77738b1492030aebf75eb/src/main/attrib.c#L142

Reply:
Based on this description of `PRINTNAME()`, my guess is that their reasoning is that this directly gives you the name node of the symbol, without any duplication, so the marking as shared makes it so that you can't accidentally overwrite that? Wouldn't it make more sense for the underlying CHARSXP node to be the thing that is marked as shared? I'm not sure

https://github.com/wch/r-source/blob/355079cacf64365cb0e1e785a9051a4660be273d/doc/manual/R-ints.texi#L433-L438

Reply:
`PRINTNAME()` returns the `CHARSXP` wrapped by the symbol. I don't think the contents of the character vectors explain the decision to mark it as shared.

The only explanation I can find is that it prevents the caller from thinking they can mutate the names and have it reflected in the pairlist.

Reply:
I've left it in for now, but feel free to remove